lmp/model/_transformer.py

```diff
@@ -0,0 +1,460 @@
+        (If you want 2 layers, then you need to set ``n_pre_hid_lyr = 1``.)
+        Must be bigger than or equal to ``1``.
+    p_emb: float
+        Dropout probability for token embeddings.
+        Must satisfy ``0.0 <= p_emb <= 1.0``.
+    p_hid: float
+        Dropout probability for every hidden representations.
+        Must satisfy ``0.0 <= p_hid <= 1.0``.
+    max_seq_len: int
+        Max length of the input sequence.
+        Must be bigger than or equal to ``1``.
+    tknzr: lmp.tknzr.BaseTknzr
+        Tokenizer instance with attributes ``pad_tkid`` and ``vocab_size``.
+    n_head: int
+        Head number of multihead attention.
+        ``d_hid`` should be divisible by it.
+
+    Attributes
+    ==========
+    emb: torch.nn.Embedding
+        Token embedding lookup matrix.
+        Use token ids to lookup token embeddings.
+    emb_dp: torch.nn.Dropout
+        Token embedding dropout.
+        Drop embedding features with probability ``p_emb``.
+    model_name: ClassVar[str]
+        Model name is ``Transformer``.
+        Used for command line argument parsing.
+    pad_tkid: int
+        Padding token id.
+        Used to create attention mask on padding tokens.
+    post_hid: torch.nn.Sequential
+        Rectified MLP which transform temporal features from hidden dimension
+        ``d_hid`` to embedding dimension ``d_emb``.
+        Drop rectified units with probability ``p_hid``.
+    pre_hid: torch.nn.Sequential
+        Rectified MLP which transform token embeddings from embedding
+        dimension ``d_emb`` to hidden dimension ``d_hid``.
+        Drop rectified units with probability ``p_hid``.
+    pe: lmp.model.PositionalEncoding
+        Positional Encoding.
+    encoderlayer: torch.nn.TransformerEncoderLayer
+        TransformerEncoderLayer is made up of MultiHeadAttention layer
+        and Feedforward layer.
+    tranformerencoder: torch.nn.TransformerEncoder
+        Stack of N encoderlayers.
+
+    References
+    ==========
+        Ashish Vaswani, Noam Shazeer, Niki Parmar, Jakob Uszkoreit,
+        Llion Jones, Aidan N. Gomez, Lukasz Kaiser, Illia Polosukhin.
+        2017. Attention is all you need.
+    """
+    model_name: ClassVar[str] = 'Transformer'
+
+    def __init__(
+            self,
+            *,
+            n_head: int,
+            d_emb: int,
+            d_hid: int,
+            n_hid_lyr: int,
+            n_post_hid_lyr: int,
+            n_pre_hid_lyr: int,
+            p_emb: float,
+            p_hid: float,
+            max_seq_len: int,
+            tknzr: BaseTknzr,
+            **kwargs: Optional[Dict],
+    ):
+        super().__init__(
+            d_emb=d_emb,
+            d_hid=d_hid,
+            n_hid_lyr=n_hid_lyr,
+            n_post_hid_lyr=n_post_hid_lyr,
+            n_pre_hid_lyr=n_pre_hid_lyr,
+            p_emb=p_emb,
+            p_hid=p_hid,
+            tknzr=tknzr,
+            **kwargs,
+        )
+
+        if d_hid % n_head != 0:
```

Comment:
Enforce type check first for parameters (including `n_head`, `max_seq_len`, etc)ㄡ